### PR TITLE
feat: extend `Graph.flix`

### PR DESCRIPTION
**Pull request by JonathanStarup**
Related to #3906 
* flipEdges
* invert
* stableDistances (handles negative cycles)
* topologicalSort
* degrees
* frontiersFrom

Missing tests

### CODE REVIEW

**Comment by magnus-madsen** [comment]
Some errors.

**Comment by JonathanStarup** [comment]
@magnus-madsen RFR. I caught a couple errors

File: main/src/library/Graph.flix [hunk 1]
```diff
@@ -108,6 +139,49 @@ namespace Graph {
     /// Returns the shortest distance between all pairs of vertices in the
     /// weighted directed graph `g`.
     ///
+    pub def stableDistances(g: m[(t, Int32, t)]): Option[Map[(t, t), Int32]] with Foldable[m], Boxable[t] = {
```
Comment (magnus-madsen) [inline main/src/library/Graph.flix:142]: Where does the name `stableDistances` come from?

File: main/src/library/Graph.flix [hunk 2]
```diff
@@ -108,6 +139,49 @@ namespace Graph {
     /// Returns the shortest distance between all pairs of vertices in the
     /// weighted directed graph `g`.
```
Comment (magnus-madsen) [inline main/src/library/Graph.flix:140]: Returns `None` if ...?

File: main/src/library/Graph.flix [hunk 3]
```diff
@@ -108,6 +139,49 @@ namespace Graph {
...
+            // nodes less than)
+            Nodes(;Set#{x}) :- Node(x).
+            NodeCount(Set.size(ns)) :- fix Nodes(;ns).
+            NodeRank(x, rank(x, ns)) :- Node(x), fix Nodes(;ns).
+            Dist(x, x, 0; Down(0)) :- Node(x).
+            Dist(x, y, 0; Down(w)) :- Edge(x, w, y).
+            Dist(x, y, k+1; w) :- Dist(x, y, k; w), NodeCount(c), if k < c.
+            Dist(x, y, k+1; w1 + w2) :- Dist(x, kth, k; w1), Dist(kth, y, k; w2),
+                NodeRank(kth, k), NodeCount(c), if k < c.
+            NegCycle() :- fix Dist(x, x, _; w), if destructDown(w) < 0.
```
Comment (magnus-madsen) [inline main/src/library/Graph.flix:159]: I worry about correctness. How small can `w` get? Can it underflow?

File: main/src/library/Graph.flix [hunk 4]
```diff
@@ -108,6 +139,49 @@ namespace Graph {
...
+            Node(x) :- Edge(_, _, x).
+            // Collect the nodes to find the rank of each node (the number of
+            // nodes less than)
+            Nodes(;Set#{x}) :- Node(x).
+            NodeCount(Set.size(ns)) :- fix Nodes(;ns).
+            NodeRank(x, rank(x, ns)) :- Node(x), fix Nodes(;ns).
+            Dist(x, x, 0; Down(0)) :- Node(x).
+            Dist(x, y, 0; Down(w)) :- Edge(x, w, y).
+            Dist(x, y, k+1; w) :- Dist(x, y, k; w), NodeCount(c), if k < c.
+            Dist(x, y, k+1; w1 + w2) :- Dist(x, kth, k; w1), Dist(kth, y, k; w2),
```
Comment (magnus-madsen) [inline main/src/library/Graph.flix:157]: I must admit that I simply dont understand this :)

File: main/src/library/Graph.flix [hunk 5]
```diff
@@ -160,19 +234,173 @@ namespace Graph {
...
+        g |> Foldable.toSet
+          |> Set.flatMap(match (a, w, b) -> Set#{(a, w, b), (b, w, a)})
+    }
+
+    ///
+    /// Returns the nodes that are at most `limit` (inclusive) distance away
+    /// from `src` in `g`.
+    /// Negative `limit` is interpreted as `0`.
+    ///
+    pub def withinDistanceOf(src: t, limit: Int32, g: m[(t, Int32, t)]): Set[t] with Foldable[m], Boxable[t] = {
```
Comment (magnus-madsen) [inline main/src/library/Graph.flix:255]: Neat!

**Review by magnus-madsen — COMMENTED** [review]
I still need to review the last functions.

File: main/src/library/Graph.flix (hunk 1 above)
Comment (JonathanStarup) [reply main/src/library/Graph.flix:142]: Oh yeah that's proabably not the best name. It just doesn't loop forever on negative cycles. You can compare it to the implementation in the examples folder

File: main/src/library/Graph.flix (hunk 3 above)
Comment (JonathanStarup) [reply main/src/library/Graph.flix:159]: It is bounded by `Dist(_, _, k; _)`. You can at most take n steps. It overflows if there is any shortest path for which the distance overflows. This is the same as distancesFrom and those related ones

File: main/src/library/Graph.flix (hunk 4 above)
Comment (JonathanStarup) [reply main/src/library/Graph.flix:157]: I wasn't too generous with the comments yeah. `Dist` is the `ShortestPath` from [wiki](https://en.wikipedia.org/wiki/Floyd%E2%80%93Warshall_algorithm) but I can take a pass with more explanation in comments

**Comment by magnus-madsen** [comment]
Do you want to take one more pass over the code and then we can merge it?

**Comment by magnus-madsen** [comment]
Compiles, but now some tests fail :)

**Comment by JonathanStarup** [comment]
RFR @magnus-madsen

**Comment by magnus-madsen** [comment]
Thanks!

**Pull request merged**